[WIP] Begin to distinguish between Perl warnings and errors

oalders opened the pull request:
This is a work in progress, but I was interested in getting some feedback.

The biggest change here is, I think, not clobbering messages from previous line numbers.  I could also change this to concatenate the messages rather than letting the first one win.

w0rp commented on ale_linters/perl/perl.vim:
```diff
@@ -27,12 +27,20 @@ function! ale_linters#perl#perl#Handle(buffer, lines) abort
+        let type = 'W'
+    endif
+
+    let seen = {}
```
You should install `Vader` for checking your Vim code. Check out the output from the Travis CI build for the warnings it will give you. It will complain about your variable names.

w0rp commented on ale_linters/perl/perl.vim:
```diff
@@ -27,12 +27,20 @@ function! ale_linters#perl#perl#Handle(buffer, lines) abort
     let l:basename = expand('#' . a:buffer . ':t')
 
+    let type = 'E'
+    let syntax_ok = matchstr(a:lines[-1], 'syntax OK')
```
If you just want to check if there is a match, you can use the operator for that. See `:help =~#`.

w0rp replied:
I also prohibit the use of `==#` and `==?` from the codebase, as they behave poorly when you mix numbers and strings. `is#` and is?` are preferred instead.

oalders replied:
Fixed in f66d38bab7

w0rp commented on ale_linters/perl/perl.vim:
```diff
@@ -27,12 +27,20 @@ function! ale_linters#perl#perl#Handle(buffer, lines) abort
+        let l:type = type
 
-        if ale#path#IsBufferPath(a:buffer, l:match[2])
+        if ( ale#path#IsBufferPath(a:buffer, l:match[2]) && !has_key(seen,l:line) )
```
You don't need the parentheses here. There's already another condition on the line below, so you could maybe move your second check to a new line below.

oalders replied:
Fixed in 43782df621315ad

oalders commented:
I think I've addressed all of the issues cited above.  I'd like to live with this for a few days before I remove the WIP label.  If this looks ok, would you like the commits squashed?

oalders commented:
BTW, I had no problems running `./run-tests` from my Mac.  Worked like a charm.

w0rp commented:
Cheers! :beers:

oalders commented:
Thanks @w0rp!
